pkg/controller/daemon/manager.go

```diff
@@ -0,0 +1,475 @@
+			// If daemon pod is supposed to be running on node, but isn't, create daemon pod.
+			nodesNeedingDaemons = append(nodesNeedingDaemons, nodeName)
+		} else if shouldRun && len(daemonPods) > 1 {
+			// If daemon pod is supposed to be running on node, but more than 1 daemon pod is running, delete the excess daemon pods.
+			// TODO: sort the daemon pods by creation time, so the the oldest is preserved.
+			for i := 1; i < len(daemonPods); i++ {
+				podsToDelete = append(podsToDelete, daemonPods[i].Name)
+			}
+		} else if !shouldRun && isRunning {
+			// If daemon pod isn't supposed to run on node, but it is, delete all daemon pods on node.
+			for i := range daemonPods {
+				podsToDelete = append(podsToDelete, daemonPods[i].Name)
+			}
+		}
+	}
+
+	// We need to set expectations before creating/deleting pods to avoid race conditions.
+	dcKey, err := controller.KeyFunc(dc)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", dc, err)
+		return
+	}
+	dm.expectations.SetExpectations(dcKey, len(nodesNeedingDaemons), len(podsToDelete))
+
+	glog.V(4).Infof("Nodes needing daemons for daemon %s: %+v", dc.Name, nodesNeedingDaemons)
+	for i := range nodesNeedingDaemons {
+		if err := dm.podControl.CreateReplicaOnNode(dc.Namespace, dc, nodesNeedingDaemons[i]); err != nil {
+			glog.V(2).Infof("Failed creation, decrementing expectations for controller %q/%q", dc.Namespace, dc.Name)
+			dm.expectations.CreationObserved(dcKey)
+			util.HandleError(err)
+		}
+	}
+
+	glog.V(4).Infof("Pods to delete for daemon %s: %+v", dc.Name, podsToDelete)
+	for i := range podsToDelete {
+		if err := dm.podControl.DeletePod(dc.Namespace, podsToDelete[i]); err != nil {
+			glog.V(2).Infof("Failed deletion, decrementing expectations for controller %q/%q", dc.Namespace, dc.Name)
+			dm.expectations.DeletionObserved(dcKey)
+			util.HandleError(err)
+		}
+	}
+}
+
+func storeDaemonStatus(dcClient client.DaemonInterface, dc *expapi.Daemon, desiredNumberScheduled, currentNumberScheduled, numberMisscheduled int) error {
+	if dc.Status.DesiredNumberScheduled == desiredNumberScheduled && dc.Status.CurrentNumberScheduled == currentNumberScheduled && dc.Status.NumberMisscheduled == numberMisscheduled {
+		return nil
+	}
+
+	var updateErr, getErr error
+	for i := 0; i <= StatusUpdateRetries; i++ {
+		dc.Status.DesiredNumberScheduled = desiredNumberScheduled
+		dc.Status.CurrentNumberScheduled = currentNumberScheduled
+		dc.Status.NumberMisscheduled = numberMisscheduled
+		_, updateErr := dcClient.Update(dc)
+		if updateErr == nil {
+			return updateErr
+		}
+		// Update the controller with the latest resource version for the next poll
+		if dc, getErr = dcClient.Get(dc.Name); getErr != nil {
+			// If the GET fails we can't trust status.Replicas anymore. This error
+			// is bound to be more interesting than the update failure.
+			return getErr
+		}
+	}
+	// Failed 2 updates one of which was with the latest controller, return the update error
```

Comment:
Should this be "StatusUpdateRetries" instead of "2"?
Also how do you kno wit is really the latest controller -- can't you always have a conflict if some other component is mutating "dc" (perhaps due to a bug)?